index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/client/EsDocument.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * ModeShape (http://www.modeshape.org)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.modeshape.jcr.index.elasticsearch.client;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import org.infinispan.schematic.DocumentFactory;
+import org.infinispan.schematic.document.Document;
+import org.infinispan.schematic.document.EditableDocument;
+import org.infinispan.schematic.document.Json;
+import org.infinispan.schematic.internal.document.BasicArray;
+
+/**
+ * Json document used for exchange data with Elasticsearch engine.
+ * 
+ * @author kulikov
+ */
+public class EsDocument {
+    private EditableDocument document = DocumentFactory.newDocument();
+    
+    /**
+     * Creates new document.
+     */
+    public EsDocument() {        
+    }
+
+    /**
+     * Creates new document with initial content.
+     * 
+     * @param origin initial content.
+     */
+    public EsDocument(Document origin) {        
+        document = DocumentFactory.newDocument(origin);
+    }
+    
+    /**
+     * Adds single property value.
+     * 
+     * @param name property name.
+     * @param value property value.
+     */
+    public void put(String name, Object value) {
+        if (value instanceof EsDocument) {
+            document.setDocument(name, ((EsDocument)value).document);
+        } else {
+            document.set(name, value);
+        }
+    }
+
+    /**
+     * Adds multivalued property value.
+     * 
+     * @param name property name.
+     * @param value property values.
+     */
+    public void put(String name, Object[] values) {
+        if (values instanceof EsDocument[]) {
+            Object[] docs = new Object[values.length];
+            for (int i = 0; i < docs.length; i++) {
+                docs[i] = ((EsDocument)values[i]).document;
+            }
+            document.setArray(name, docs);
+        } else {
+            document.setArray(name, values);
+        }
+    }
+    
+    /**
+     * Gets property value.
+     * 
+     * @param name property name.
+     * @return property value.
+     */
+    public Object get(String name) {
+        Object obj = document.get(name);
+        if (obj instanceof BasicArray) {
+           Object[] res = new Object[((BasicArray)obj).size()];
```

Comment:
Perhaps `Array` (which `BasicArray` implements) also implements `List<Object>`, so it should be possible to just call `array.toArray(new Object[array.size()])`. Perhaps `Array` should provide a more convenient method for doing exactly this.

Although, I still think that if the ES index provider code directly used `Document` and `Array`, this would not really be needed.